tests/unit/src/test/scala/tests/codeactions/FlatMapToForComprehensionSuite.scala

```diff
@@ -0,0 +1,142 @@
+package tests.codeactions
+
+import scala.meta.internal.metals.codeactions.FlatMapToForComprehensionCodeAction
+import scala.meta.internal.metals.codeactions.RewriteBracesParensCodeAction
+
+class FlatMapToForComprehensionSuite
+    extends BaseCodeActionLspSuite("forComprehension") {
+
+  check(
+    "mixture",
+    """|object A {
+       |    def double(x : Int, y: Int = 1) = y * x
+       |    def check(x: Int) = true
+       |    val list = List(1, 2, 3)
+       |    def negate(a: Boolean) = !a
+       |
+       |    val res3 = list
+       |        .flatMap{
+       |              a =>
+       |                      {
+       |                        val m = 6
+       |                        Some(a + 1).map(b => b + 3 + 4)
+       |                      }
+       |
+       |        }.filter (check)
+       |        .filterNot(_ =>  true)
+       |        .map(_ => 7)
+       |        .map(c => c - 1)
+       |        .map(5 + double(_, 7).toFloat.toInt / 8 + 6)
+       |        .filter(d => d > 1)
+       |        .map(double(_, 5))
+       |        .map( double(_, 4).toFloat.toDouble)
+       |        .map( _.toInt.compare(3))
+       |        .map(_ > 2)
+       |        .map(!negate(_))
+       |        .m<<>>ap( true && !negate(_) && false)
```

Comment:
Also, could we make the full for comprehension change even if we don't put the cursor on the last map?

Reply:
No, we should not, and we don’t. We only do the for comprehension conversion for the `Term.Apply` that is captured by the cursor, which is the correct and expected behavior. so let’s say, if someone puts the cursor on the `filter` in the middle, the conversion is only going to happen over this, and the result will be wrapped in braces.

```
list
        |        .flatMap{
        |              a =>
        |                      {
        |                        val m = 6
        |                        Some(a + 1).map(b => b + 3 + 4)
        |                      }
        |
        |        }.filter (check)
        |        .filterNot(_ =>  true)
        |        .map(_ => 7)
        |        .map(c => c - 1)
        |        .map(5 + double(_, 7).toFloat.toInt / 8 + 6)
        |        .filter(d => d > 1)

```

Reply:
One scenario in which the user might want to have control over which part of the chain gets converted to for comprehension, happens when one of the maps, has a body of say 10 lines of code. Putting this map in the enumerators of the for comprehension, makes the code less readable, so the user might want to have the option of only converting what’s above it, or put that map in the yield part.

Reply:
> One scenario in which the user might want to have control over which part of the chain gets converted to for comprehension, happens when one of the maps, has a body of say 10 lines of code. 

I don't think we should encourage code that only converts parts of the code. We should always convert everything, it's not intuitive that the user needs to have the cursor on a specific place. We need to convert the whole chain.

Reply:
https://user-images.githubusercontent.com/93259201/173059776-697fec5c-9102-4567-907e-1cc0e6bec759.mov

What do you think?

Reply:
I would honestly rewrite the whole thing unless someone at some point reports the need for partial rewrite. I can't imagine a scenario that user would want to rewrite only part into a for comprehension.

Reply:
No problem, then🙃